filament/src/details/UboManager.cpp

```diff
@@ -201,69 +187,76 @@ void UboManager::updateSlot(DriverApi& driver, AllocationId id,
     driver.copyToMemoryMappedBuffer(mMemoryMappedBufferHandle, offset, std::move(bufferDescriptor));
 }
 
-void UboManager::retireSlot(BufferAllocator::AllocationId id) {
+void UboManager::manageMaterialInstance(FMaterialInstance* instance) {
+    mPendingInstances.insert(instance);
+}
+
+void UboManager::unmanageMaterialInstance(const FMaterialInstance* materialInstance) {
+    AllocationId id = materialInstance->getAllocationId();
     if (!BufferAllocator::isValid(id))
         return;
-    mAllocator.retire(id);
-}
 
-UboManager::AllocationResult UboManager::allocateOnDemand(
-        const std::unordered_map<const FMaterial*, ResourceList<FMaterialInstance>>&
-                materialInstances) {
-    // Collect all MIs that need allocation into two groups.
-    std::vector<FMaterialInstance*> newInstances;
-    std::vector<FMaterialInstance*> existingInstances;
-    for (const auto& [_, miList] : materialInstances) {
-        miList.forEach([&](FMaterialInstance* mi) {
-            if (!mi->isUsingUboBatching()) {
-                return;
-            }
-            if (BufferAllocator::isValid(mi->getAllocationId())) {
-                existingInstances.push_back(mi);
-            } else {
-                newInstances.push_back(mi);
-            }
-        });
+    // const_cast should be safe here since this cast is just to match the container type.
+    auto mi = const_cast<FMaterialInstance*>(materialInstance);
+    if (UTILS_UNLIKELY(mPendingInstances.contains(mi))) {
+        mPendingInstances.erase(mi);
+    }
```

Comment:
do we need to do the `if`? is it not okay to just to erase(mi), if it's not contained, it won't erase it!